cpp-package/src/inc/brainflow_array.h

```diff
@@ -0,0 +1,468 @@
+        , origin (nullptr)
+    {
+        static_assert (Dim == 3, "This function is only for BrainFlowArray<T, 3>");
+        origin = new T[size0 * size1 * size2];
+        memcpy (origin, ptr, size0 * size1 * size2 * sizeof (T));
+    }
+
+    explicit BrainFlowArray (const BrainFlowArray &other)
+        : length (other.length), size (other.size), stride (other.stride), origin (nullptr)
+    {
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+    }
+
+    const BrainFlowArray &operator= (const BrainFlowArray &other)
+    {
+        length = other.length;
+        size = other.size;
+        stride = other.stride;
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+
+        return *this;
+    }
+
+    BrainFlowArray (BrainFlowArray &&other)
+        : length (other.length)
+        , size (std::move (other.size))
+        , stride (std::move (other.stride))
+        , origin (nullptr)
+    {
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+
+        other.length = 0;
+        other.size = zero_array<Dim> ();
+        other.origin = nullptr;
+    }
+
+    const BrainFlowArray &operator= (BrainFlowArray &&other)
+    {
+        length = other.length;
+        size = std::move (other.size);
+        stride = std::move (other.stride);
+        origin = new T[other.length];
+        memcpy (origin, other.origin, sizeof (T) * other.length);
+
+        other.length = 0;
+        other.size = zero_array<Dim> ();
+        other.origin = nullptr;
+
+        return *this;
+    }
+
+    /// get size of data type
+    int get_item_size () const
+    {
+        return sizeof (T);
+    }
+
+    /// get total length
+    int get_length () const
+    {
+        return length;
+    }
+
+    /// get ndims
+    int get_ndims () const
+    {
+        return Dim;
+    }
+
+    /// get dims size
+    const std::array<int, Dim> &get_sizes () const
+    {
+        return size;
+    }
+
+    /// get size in dim
+    int get_size (int dim) const
+    {
+        if ((dim < 0) || (dim >= Dim))
+        {
+            throw BrainFlowException (
+                "invalid dim argument", (int)BrainFlowExitCodes::INVALID_ARGUMENTS_ERROR);
+        }
+        return size[dim];
+    }
+
+    const std::array<int, Dim> &get_strides () const
+    {
+        return stride;
+    }
+
+    int get_stride (int dim) const
+    {
+        return stride[dim];
+    }
+
+    /// get raw pointer
+    T *get_raw_ptr ()
+    {
+        return origin;
+    }
+
+    /// get raw pointer
+    const T *get_raw_ptr () const
+    {
+        return origin;
+    }
+
+    operator T * ()
+    {
+        return get_raw_ptr ();
+    }
+
+    operator const T * () const
+    {
+        return get_raw_ptr ();
+    }
+
+    /// access element at index
+    T &at (int index0)
```

Comment:
C++ developers usually distinguish between `at()` and `operator[]` in that `at()` throws an out of bounds exception whereas `operator[]` does not check bounds.  Here you do not throw an exception which would be misleading to me, but I don't know whether the same paradigm holds true for researchers as opposed to developers.